docs/data/material/customization/theme-components/theme-components.md

```diff
@@ -56,48 +56,114 @@ Each component is composed of several different parts.
 These parts correspond to classes that are available to the component—see the **CSS** section of the component's API page for a detailed list.
 You can use these classes inside the `styleOverrides` key to modify the corresponding parts of the component.
 
+### Overrides based on props
+
+Each slot of the component's `styleOverrides` support a key of `variants`. Variants are sets of overrides when the specified props are matched.
+
+The definitions are specified in an array with sensitive order, so make sure that the styles that should win are specified last.
+
 ```js
 const theme = createTheme({
   components: {
     MuiButton: {
       styleOverrides: {
-        root: ({ ownerState }) => ({
-          ...(ownerState.variant === 'contained' &&
-            ownerState.color === 'primary' && {
-              backgroundColor: '#202020',
-              color: '#fff',
-            }),
-        }),
+        root: {
+          variants: [
+            {
+              props: { variant: 'dashed' },
+              style: {
+                textTransform: 'none',
+                border: `2px dashed ${blue[500]}`,
+              },
+            },
+            {
+              props: { variant: 'dashed', color: 'secondary' },
+              style: {
+                border: `4px dashed ${red[500]}`,
+              },
+            },
+          ],
+        },
       },
     },
   },
 });
 ```
 
-### Overrides based on props
+If you're using TypeScript, you'll need to specify your new variants/colors, using [module augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation).
```

Comment:
I would recommend having two separate sections as before, so it's easier to search for custom variants for example. If people are not looking for custom variant, it is surprising why they would see variant 'dashed' in this demo.

Reply:
To be honest, I am opposed to that idea, that's why I removed it.

"Overrides based on props" is what users want, "Creating new variants" is our technical term.

Users should learn that if they want to override CSS based on props, they should use `variants` but not the other way around.

However, adding `variants` to the title could help users that already familiar with the word find this section faster. So how about changing it to `Overrides based on props using variants`? Any thoughts @samuelsycamore?

Reply:
I added more examples instead, can you check them first.

Reply:
> "Overrides based on props" is what users want, "Creating new variants" is our technical term.

I don't think "Creating new variants" is our technical term. The term "variants" is now standard across the industry and mainstream design tools like Figma use it. So I think "adding/removing variants" is a correct mental model for many users.

_For another PR:_ I think we should mention how to remove variants.

Reply:
> "variants" is now standard across the industry and mainstream design tools like Figma use it

Sounds reasonable to me. Then it's better to change `Overrides based on props` to `Creating new variants`.